README.md
# Form Generator Sprinkle for [UserFrosting 5](https://www.userfrosting.com)

[![Donate][kofi-badge]][kofi]
[![Latest Version][releases-badge]][releases]
[![UserFrosting Version][uf-version]][uf]
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg)](LICENSE)
[![Build][build-badge]][build]
[![PHPStan][PHPStan-badge]][PHPStan]
[![Codecov][codecov-badge]][codecov]
[![StyleCI][styleci-badge]][styleci]

[kofi]: https://ko-fi.com/A7052ICP
[kofi-badge]: https://img.shields.io/badge/Donate-Buy%20Me%20a%20Coffee-blue?logo=ko-fi&logoColor=white
[releases]: https://github.com/lcharette/UF_FormGenerator/releases
[releases-badge]: https://img.shields.io/github/release/lcharette/UF_FormGenerator.svg
[uf-version]: https://img.shields.io/badge/UserFrosting->=%205.0-brightgreen.svg
[uf]: https://github.com/userfrosting/UserFrosting
[build]: https://github.com/lcharette/UF_FormGenerator/actions?query=workflow%3ABuild
[build-badge]: https://img.shields.io/github/actions/workflow/status/lcharette/UF_FormGenerator/Build.yml?branch=5.0&logo=github
[PHPStan]: https://github.com/lcharette/UF_FormGenerator/actions?query=workflow%3APHPStan
[PHPStan-badge]: https://img.shields.io/github/actions/workflow/status/lcharette/UF_FormGenerator/PHPStan.yml?branch=5.0&label=PHPStan
[codecov]: https://codecov.io/gh/lcharette/UF_FormGenerator
[codecov-badge]: https://codecov.io/gh/lcharette/UF_FormGenerator/branch/5.0/graph/badge.svg
[styleci]: https://styleci.io/repos/68563337
[styleci-badge]: https://styleci.io/repos/68563337/shield?branch=5.0&style=flat

This Sprinkle provides helper classes, Twig template and JavaScript plugins to generate HTML forms, modals and confirm modal bases on UserFrosting [validation schemas](https://learn.userfrosting.com/routes-and-controllers/client-input/validation).

# Help and Contributing

If you need help using this sprinkle or found any bug, feels free to open an issue or submit a pull request. You can also find me on the [UserFrosting Chat](https://chat.userfrosting.com/) most of the time for direct support.

# Versions and UserFrosting support

| UserFrosting Version | FormGenerator Version |
| :------------------: | :-------------------: |
|        5.0.x         |         5.0.x         |
|        4.4.x         |         4.x.x         |
|        4.3.x         |    [3.0.x] & 4.0.x    |
|        4.2.x         |        [3.0.x]        |
|        4.1.x         |        [2.0.x]        |
|        4.0.x         |      No Support       |

[3.0.x]: https://github.com/lcharette/UF_FormGenerator/tree/3.0#form-generator-sprinkle-for-userfrosting-4
[2.0.x]: https://github.com/lcharette/UF_FormGenerator/tree/2.2#form-generator-sprinkle-for-userfrosting-4

# Installation
To install FormGenerator in your sprinkle : 

1. Install FormGenerator through Composer:
    ```
    composer require lcharette/uf_formgenerator "^5.0"
    ```

2. Add `UserFrosting\Sprinkle\FormGenerator\FormGenerator` to your Sprinkle Recipe sprinkle method. 

3. To use the frontend helper, first install the npm dependency:
    ```
    npm install --save @lcharette/formgenerator
    ```

4. Then add this entry to your `webpack.config.js`, in the `sprinkles` list : 
    ```js
    FormGenerator: require('@lcharette/formgenerator/webpack.entries'),
    ```

5. Each template file where you want to use the frontend helper, add this line to `{% block scripts_page %}`:
    ```
    {{ encore_entry_script_tags('widget.formGenerator') }}
    ```

6. Run `php bakery bake` to finish installation of the sprinkle.

Alternatively, you can add `@lcharette/formgenerator/app/assets/js/widget-formGenerator.js` globally in your app main js file.

# Working example

The `public/` directory serves as an example of FormGenerator. You can clone this repository and install as any UserFrosting 5 sprinkle :
1. `composer install`
2. `php bakery bake`
3. `php -S localhost:8080 -t public`

This demo is not linked to any database tables, so changes are not actually saved ;)

![Screenshot 1](/.github/screenshots/UF_FormGeneratorExample1.png?raw=true)
![Screenshot 2](/.github/screenshots/UF_FormGeneratorExample2.png?raw=true)
![Screenshot 3](/.github/screenshots/UF_FormGeneratorExample3.png?raw=true)
![Screenshot 4](/.github/screenshots/UF_FormGeneratorExample4.png?raw=true)

# Features and usage
Before starting with _FormGenerator_, you should read the main UserFrosting guide to familiarize yourself with _validation schemas_: (https://learn.userfrosting.com/routes-and-controllers/client-input/validation).

## Form generation
### Defining the fields in the schema
This sprinkle uses the `schemas` used by UserFrosting to validate form data to build form. To achieve this, a new `form` key is simply added to the fields found in a `schema` file.

For example, here's a simple `schema` used to validate a form used to create a `project`. The form will contain a `name`, `description` and `status` fields.

```json
{
    "name" : {
        "validators" : {
            "length" : {
                "min" : 1,
                "max" : 100
            },
            "required" : {
                "message" : "PROJECT.VALIDATE.REQUIRED_NAME"
            }
        }
    },
    "description" : {
        "validators" : {}
    },
    "status" : {
        "validators" : {
            "member_of" : {
                "values" : [
                    "0", "1"
                ]
            },
            "required" : {
                "message" : "PROJECT.VALIDATE.STATUS"
            }
        }
    }
}
```
> Note: FormGenerator works with json and YAML schemas.

At this point, with typical UserFrosting setup, you would be going into your controller and Twig files to manually create your HTML form. This can be easy if you have a two or three fields, but can be a pain with a dozen fields and more. This is where FormGenerator steps in with the use of a new `form` attribute. Let's add it to our `project` form :

```json
{
    "name" : {
        "validators" : {
            "length" : {
                "min" : 1,
                "max" : 100
            },
            "required" : {
                "message" : "VALIDATE.REQUIRED_NAME"
            }
        },
        "form" : {
            "type" : "text",
            "label" : "NAME",
            "icon" : "fa-flag",
            "placeholder" : "NAME"
        }
    },
    "description" : {
        "validators" : {},
        "form" : {
            "type" : "textarea",
            "label" : "DESCRIPTION",
            "icon" : "fa-pencil",
            "placeholder" : "DESCRIPTION",
            "rows" : 5
        }
    },
    "status" : {
        "validators" : {
            "member_of" : {
                "values" : [
                    "0", "1"
                ]
            },
            "required" : {
                "message" : "VALIDATE.STATUS"
            }
        },
        "form" : {
            "type" : "select",
            "label" : "STATUS",
            "options" : {
                "0" : "Active",
                "1" : "Disabled"
            }
        }
    }
}
```

Let's look closer at the `name` field :

```json
"form" : {
    "type" : "text",
    "label" : "PROJECT.NAME",
    "icon" : "fa-flag",
    "placeholder" : "PROJECT.NAME"
}
```

Here you can see that we define the `type`, `label`, `icon` and `placeholder` value for this `name` field. You can define any standard [form attributes](http://www.w3schools.com/html/html_form_attributes.asp), plus the `icon`, `label` and `default` attributes. `data-*` attributes can also be defined in your schema if you need them. For the `select` element, a special `options` attribute containing an array of `key : value` can be used to define the dropdown options. The select options (as any other attributes) can also be set in PHP (see further below).

And of course, the values of the `label` and `placeholder` attributes can be defined using _translation keys_.

Currently, FormGenerator supports the following form elements :
- text (and any input supported by the HTML5 standard : number, tel, password, etc.)
- textarea
- select
- checkbox
- hidden
- alert (Display a static alert box in the form)

### The controller part
Once your fields defined in the `schema` json or yaml file, you need to load that schema in your controller.

First thing to do is add FormGenerator's `Form` class to your "use" list :
`use UserFrosting\Sprinkle\FormGenerator\Form;`

Next, where you load the schema and setup the `validator`, you simply add the new Form creation:
```php
// Load validator rules
$schema = new RequestSchema("schema://project.json");
$validator = new JqueryValidationAdapter($schema, $this->translator);

// Create the form
$form = new Form($schema, $project);
```

In this example, `$project` can contain the default (or current value) of the fields. A data collection fetched from the database with eloquent can also be passed directly. That second argument can also be omitted to create an empty form.

Last thing to do is send the fields to Twig. In the list of returned variables to the template, add the `fields` variable:
```php
$view->render($response, "pages/myPage.html.twig", [
    "fields" => $form->generate(),
    "validators" => $validator->rules('json', true)
]);

```

### The Twig template part

Now it's time to display the form in `myPage.html.twig` !

```html
<form name="MyForm" method="post" action="/Path/to/Controller/Handling/Form">
    {% include "forms/csrf.html.twig" %}
    <div id="form-alerts"></div>
    <div class="row">
        <div class="col-sm-8">
            {% include 'FormGenerator/FormGenerator.html.twig' %}
        </div>
    </div>
    <div class="row">
      <button type="submit" class="btn btn-block btn-lg btn-success">Submit</button>
    </div>
</form>
```

That's it! No need to list all the field manually. The ones defined in the `fields` variable will be displayed by `FormGenerator/FormGenerator.html.twig`. Note that this will only load the fields, not the form itself. The `<form>` tag and `submit` button needs to be added manually.

## Modal form
What if you want to show a form in a modal window? Well, FormGenerator makes it even easier! It's basically three steps:
1. Setup your form schema (as described above)
2. Setup the form in your controller
3. Call the modal from your template

### Setup the form in your controller
With your schema in hand, it's time to create a controller and route to load your modal. The controller code will be like any basic UserFrosting modal, plus the `$form` part above and one changes in the `render` part. For example :

```php
$view->render($response, "FormGenerator/modal.html.twig", [
    "box_id"        => $get['box_id'],
    "box_title"     => "PROJECT.CREATE",
    "submit_button" => "CREATE",
    "form_action"   => '/project/create',
    "fields"        => $form->generate(),
    "validators"    => $validator->rules('json', true)
]);
```

As you can see, instead of rendering your own Twig template, you simply have to specify FormGenerator's modal template. This template requires the following variables:
1. `box_id`: This should always be `$get['box_id']`. This is used by the JavaScript code to actually display the modal.
2. `box_title`: The title of the modal.
3. `submit_button`: The label of the submit button. Optional. Default to `SUBMIT` (localized).
4. `form_action`: The route where the form will be sent
5. `fields`: The fields. Should always be `$form->generate()`
6. `validators`: Client side validators

### Call the modal from your template
So at this point you have a controller that displays the modal at a `/path/to/controller` route. Time to show that modal. Again, two steps:

First, define a link or a button that will call the modal when clicked. For example :
```html
<button class="btn btn-success js-displayForm" data-formUrl="/path/to/controller">Create</button>
```

The important part here is the `data-formUrl` attribute. This is the route that will load your form. `js-displayForm` is used here to bind the button to the action.

Second, load the FormGenerator JavaScript widget. Add this to your Twig file:
```
{% block scripts_page %}
    {{ assets.js('js/FormGenerator') | raw }}
{% endblock %}
```

By default, the `formGenerator` plugin will bind a **form modal** to every element with the `js-displayForm` class.

## Modal confirmation

One side features of FormGenerator is the ability to add a confirmation modal to your pages with simple HTML5 attributes. The process is similar to adding a modal form, without the need to create any controller or route.

Let's look at a delete button / confirmation for our `project` :
```html
<a href="#" class="btn btn-danger js-displayConfirm"
  data-confirm-title="Delete project ?"
  data-confirm-message="Are you sure you want to delete this project?"
  data-confirm-button="Yes, delete project"
  data-post-url="/project/delete"><i class="fa fa-trash-o"></i> Delete</a>
```
(Note that content of data attributes can be translation keys)

If not already done, make sure the FormGenerator assets are included in your template.
```
{% block scripts_page %}
    {{ assets.js('js/FormGenerator') | raw }}
{% endblock %}
```

By default, the `formGenerator` plugin will bind a **confirmation modal** to every element with the `js-displayConfirm` class.

## Advance usage

### Defining attributes in PHP

#### setInputArgument

Form field input attributes can also be added or edited from PHP. This can be useful when dynamically defining a Select input options. To do this, simply use the `setInputArgument($inputName, $property, $data)` method. For example, to add a list to a `clients` select :

```php
// Get clients from the db model
$clients = Clients::all();

$form = new Form($schema);
$form->setInputArgument('clients', 'options', $clients);
```

#### setData

If you want to set the form values once the form instance is created, you can use the `setData($data)` method:

```php
$form = new Form($schema);
$form->setData($clients, $project);
```

#### setValue

Similar to the `setData` method, you can set a specific input value using the `setValue($inputName, $value)` method :

```php
$currentClient = ...

$form = new Form($schema, $project);
$form->setValue('clients', $currentClient);
```

#### setFormNamespace

When dealing with multiple form on the same page or a dynamic number of input (you can use the new `Loader` system in 4.1 to build dynamic schemas!), it can be useful to wrap form elements in an array using the `setFormNamespace($namespace)` method. This can also your the input names [to contains dot syntaxt](http://stackoverflow.com/a/20365198/445757).

For example, `$form->setFormNamespace("data");` will transform all the input names from `<input name="foo" [...] />` to `<input name="data[foo]" [...] />`.

#### registerType

If you want to overwrite or add a new element type,


First, you need to create the element itself. This class needs to extends the `UserFrosting\Sprinkle\FormGenerator\Element\Input` class. In there you can define the default attributes, and do other transformation. For example, to define a new `Date` element type :

```php
<?php

namespace UserFrosting\Sprinkle\MySprinkle\Element;

use UserFrosting\Sprinkle\FormGenerator\Element\Input;

class Date extends Input
{
    protected function applyTransformations(): void
    {
        $this->element = array_merge([
            'class' => 'myDateElement',
            'value' => $this->getValue(),
            'name'  => $this->name,
            'id'    => 'field_' . $this->name,
            'date-foo' => //...
        ], $this->element);
    }
}
```

Next, you need to register your `Date` element type. If the `date` type is already registered, it will be overwritten by your custom class.

```php
$form = new Form($schema, $project);
$form->registerType('date', UserFrosting\Sprinkle\MySprinkle\Element\Date::class);
```

### Javascript Plugin

By default, the `formGenerator` plugin will bind a **form modal** to every element with the `js-displayForm` class and will bind a **confirmation modal** to every element with the `js-displayConfirm` class. You can

#### Options
The following options are available:

Just pass an object with those
 - `mainAlertElement` (jQuery element). The element on the main page where the main alerts will be displayed. Default to `$('#alerts-page')`.
 - `redirectAfterSuccess` (bool). If set to true, the page will reload when the form submission or confirmation is successful. Default to `true`.

Example:
```js
$(".project-edit-button").formGenerator({redirectAfterSuccess: false});
```

#### Events
You can listen for some events returned by FormGenerator. Those events can be used to apply some actions when the modal is displayed or the form is successfully sent. For example, this is can be used with `redirectAfterSuccess` on `false` to refresh the data on the page when the form is submitted successfully.

- `formSuccess.formGenerator`
- `displayForm.formGenerator`
- `displayConfirmation.formGenerator`
- `confirmSuccess.formGenerator`
- `error.formGenerator`

Example:
```js
$(".project-edit-button").on("formSuccess.formGenerator", function () {
    // Refresh data
});
```

# Running tests

FormGenerator comes with some unit tests. Before submitting a new Pull Request, you need to make sure all tests are a go. With the sprinkle added to your UserFrosting installation, simply execute the `php bakery test FormGenerator` command to run the tests.

# License

By [Louis Charette](https://bbqsoftwares.com). Copyright (c) 2020, free to use in personal and commercial software as per the MIT license.


CHANGELOG.md
# Change Log

## [5.0.0](https://github.com/lcharette/UF_FormGenerator/compare/4.0.1...5.0.0)
- Added support for UserFrosting 5
- Data passed to forms can be mixed type, not only string. Data will still be casted to string. 
- Merge [#12](https://github.com/lcharette/UF_FormGenerator/pull/12) - Remove fa class to allow for FA v5 fas & fab - Thanks [@phillmac](https://github.com/phillmac)

## [4.0.1]
- Fixed issue where the value of an element whose name is using dot notation wouldn't be found in the corresponding multidimensional array.

## [4.0.0]
- Form nows register element class, instead of using hardcoded string.
- Form has new `registerType` method to register new custom elements. `getType(string $name)` and `getTypes()` can be used to get the class for a type (eg. 'text', 'select', etc.) abd a list of types. `removeType` will unregister a type.
- Form still fallback to `Text` type if type is not found, but this can now be changed with `getDefaultType` and `setDefaultType`.
- Form uses `RequestSchemaRepository` instead of `RequestSchemaInterface`.
- Stricter PHP7 type throughout.
- Renamed `BaseInput` to `Input`.
- `Input` has new `setValue`, `getElement`, `setElement`, `getName` & `setName` methods.
- Checkbox element now pass the value to non-binary checkbox.
- `FormGeneratorController` uses strict typing.
- Finished tests and fix tests namespace. 100% coverage!
- Added PHP-CS-Fixer, PHPStan, StyleCI configuration.
- Updated Travis setup.

## [3.0.0]
- Added `successCallback` option
- Support for UserFrosting 4.2

## [2.2.10]
- Added support for Repository

## [2.2.9]
- Fix issue when setting data that is a collection
- `formSuccess` and `confirmSuccess` events now include the request data as a second argument

## [2.2.8]
- Fix icon in textarea macro

## [2.2.7]
- Added `modal-large` template file.

## [2.2.6]
- Fix issue with `binary` checkbox tests.
- Fix Text input style when no icon is added

## [2.2.5]
- Added `binary` option for checkbox to disable UF binary checkbox system (bool; default true).

## [2.2.4]
- Add necessary HTML to disable submit and cancel button in modal form.

## [2.2.3]
- New `$form->setOptions` function to set options of a select element. Shortcut for using `setInputArgument` and `setValue`.

## [2.2.2]
- Fix issue with error alert no displaying on confirmation dialog

## [2.2.1]
- Initialize ufAlert if not already done
- Autofocus first form field when modal is displayed

## [2.2.0]
- Refactored the javascript plugin
- Added new events
- Added new `redirectAfterSuccess` boolean option

## [2.1.2]
- Fix warning with select macro

## [2.1.1]
- Fix issue with the select macro
- Renamed macro templates with the `*.html.twig` extension

## [2.1.0]
- Completely refactored how form fields are parsed, including how default value are defined. Each input type now defines it's own class for defining default values and transforming some input.
- Twig templates updated to reflect the new parser.
- Twig macros changed from `*.generate(name, value)` to `*.generate(input)`.
- **`Bool` type changed to `checkbox`**.
- Removed the `number` Twig template (Will use the text input one).
- Added unit tests.
- Support for any attributes in the schema. For example, if you need to add a data attribute to a field, your schema would be:
```
"myField" : {
    "form" : {
        "type" : "text",
        "label" : "My Field",
        "icon" : "fa-pencil",
        "data-something" : "blah"
    }
}
```

## [2.0.0]
- Updated for UserFrosting v4.1.x

The custom `RequestSchema` have been removed. Instead of building the form directly on the schema using `$schema->initForm()`, you now create a new Form using `$form = new Form($schema)` and go on from there. Handling complex schema can now be done using the new loader system from UF 4.1.

`$schema->generateForm();` has also been changed to `$form->generate();`.

## [1.0.1]
- Bug fixes

## 1.0.0
- Initial release

<!--
## [Unreleased]

### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security
-->

[4.0.1]: https://github.com/lcharette/UF_FormGenerator/compare/4.0.0...4.0.1
[4.0.0]: https://github.com/lcharette/UF_FormGenerator/compare/3.0.0...4.0.0
[3.0.0]: https://github.com/lcharette/UF_FormGenerator/compare/2.2.10...3.0.0
[2.2.10]: https://github.com/lcharette/UF_FormGenerator/compare/2.2.9...2.2.10
[2.2.9]: https://github.com/lcharette/UF_FormGenerator/compare/2.2.8...2.2.9
[2.2.8]: https://github.com/lcharette/UF_FormGenerator/compare/2.2.7...2.2.8
[2.2.7]: https://github.com/lcharette/UF_FormGenerator/compare/2.2.6...2.2.7
[2.2.6]: https://github.com/lcharette/UF_FormGenerator/compare/2.2.5...2.2.6
[2.2.5]: https://github.com/lcharette/UF_FormGenerator/compare/2.2.4...2.2.5
[2.2.4]: https://github.com/lcharette/UF_FormGenerator/compare/2.2.3...2.2.4
[2.2.3]: https://github.com/lcharette/UF_FormGenerator/compare/2.2.2...2.2.3
[2.2.2]: https://github.com/lcharette/UF_FormGenerator/compare/2.2.1...2.2.2
[2.2.1]: https://github.com/lcharette/UF_FormGenerator/compare/2.2.0...2.2.1
[2.2.0]: https://github.com/lcharette/UF_FormGenerator/compare/2.1.2...2.2.0
[2.1.2]: https://github.com/lcharette/UF_FormGenerator/compare/2.1.1...2.1.2
[2.1.1]: https://github.com/lcharette/UF_FormGenerator/compare/2.1.0...2.1.1
[2.1.0]: https://github.com/lcharette/UF_FormGenerator/compare/2.0.0...2.1.0
[2.0.0]: https://github.com/lcharette/UF_FormGenerator/compare/1.0.1...2.0.0
[1.0.1]: https://github.com/lcharette/UF_FormGenerator/compare/1.0.0...1.0.1


app/src/Form.php
<?php

declare(strict_types=1);

/*
 * UserFrosting Form Generator
 *
 * @link      https://github.com/lcharette/UF_FormGenerator
 * @copyright Copyright (c) 2020 Louis Charette
 * @license   https://github.com/lcharette/UF_FormGenerator/blob/master/LICENSE (MIT License)
 */

namespace UserFrosting\Sprinkle\FormGenerator;

use Illuminate\Contracts\Config\Repository;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Support\Arr;
use Illuminate\Support\Collection;
use InvalidArgumentException;
use UserFrosting\Fortress\RequestSchema\RequestSchemaRepository;
use UserFrosting\Sprinkle\FormGenerator\Element\Alert;
use UserFrosting\Sprinkle\FormGenerator\Element\Checkbox;
use UserFrosting\Sprinkle\FormGenerator\Element\Hidden;
use UserFrosting\Sprinkle\FormGenerator\Element\InputInterface;
use UserFrosting\Sprinkle\FormGenerator\Element\Select;
use UserFrosting\Sprinkle\FormGenerator\Element\Text;
use UserFrosting\Sprinkle\FormGenerator\Element\Textarea;
use UserFrosting\Sprinkle\FormGenerator\Exception\ClassNotFoundException;
use UserFrosting\Sprinkle\FormGenerator\Exception\InputNotFoundException;
use UserFrosting\Sprinkle\FormGenerator\Exception\InvalidClassException;

/**
 * Form Class.
 *
 * The FormGenerator class, which is used to process the `form` part from a Fortress
 * schema into an html form element for Twig.
 */
class Form
{
    /**
     * @var RequestSchemaRepository The form fields definition
     */
    protected $schema;

    /**
     * @var array<string,string|int> The form values
     */
    protected $data = [];

    /**
     * @var string Used to wrap form fields in top-level array
     */
    protected $formNamespace = '';

    /**
     * @var array<string,string> List of input type classes registered
     */
    protected $types = [];

    /**
     * @var string Default input type for element without one.
     */
    protected $defaultType = 'text';

    /**
     * Class constructor.
     *
     * @param RequestSchemaRepository                          $schema
     * @param array<string>|Collection<mixed>|Model|Repository $data   (default: [])
     */
    public function __construct(RequestSchemaRepository $schema, $data = [])
    {
        $this->setSchema($schema);
        $this->setData($data);

        // Register default types
        $this->registerDefaultType();
    }

    /**
     * Set the form current values.
     *
     * @param array<string>|Collection<mixed>|Model|Repository $data The form values
     *
     * @return self
     */
    public function setData($data)
    {
        if ($data instanceof Collection || $data instanceof Model) {
            $this->data = $data->toArray();
        } elseif ($data instanceof Repository) {
            $this->data = $data->all();
        } elseif (is_array($data)) {
            $this->data = $data;
        } else {
            throw new InvalidArgumentException('Data must be an array, a Collection, a Model or a Repository');
        }

        return $this;
    }

    /**
     * Set the schema for this validator.
     *
     * @param RequestSchemaRepository $schema A RequestSchemaRepository object, containing the form definition.
     *
     * @return self
     */
    public function setSchema(RequestSchemaRepository $schema)
    {
        $this->schema = $schema;

        return $this;
    }

    /**
     * Get all registered type classes.
     *
     * @return array<string,string>
     */
    public function getTypes(): array
    {
        return $this->types;
    }

    /**
     * Get the class for a specific name.
     *
     * @throws InputNotFoundException if type name not defined
     *
     * @return string The type name to return
     */
    public function getType(string $name): string
    {
        if (!isset($this->types[$name])) {
            throw new InputNotFoundException("Input type $name not found");
        }

        return $this->types[$name];
    }

    /**
     * Register a custom input type class.
     * Will overwrite any previously defined clas with the same name.
     *
     * @param string $name  The type name (eg. select, text, phone, mytext, etc.)
     * @param string $class The class uses
     */
    public function registerType(string $name, string $class): void
    {
        if (!class_exists($class)) {
            throw new ClassNotFoundException("Class $class not found.");
        }

        $this->types[$name] = $class;
    }

    /**
     * Remove input type from registered list.
     *
     * @param string $name
     */
    public function removeType(string $name): void
    {
        unset($this->types[$name]);
    }

    /**
     * Define the value of a specific form input.
     *
     * @param string     $inputName The input name
     * @param string|int $value     Form value
     *
     * @return self
     */
    public function setValue(string $inputName, $value)
    {
        $this->data[$inputName] = $value;

        return $this;
    }

    /**
     * Function used to overwrite the input argument from a schema file.
     * Can also be used to overwrite an argument hardcoded in the Twig file.
     *
     * @param string $inputName The input name where the argument will be added
     * @param string $property  The argument name. Example "data-color"
     * @param mixed  $value     The value of the argument
     *
     * @return self
     */
    public function setInputArgument(string $inputName, string $property, $value)
    {
        if ($this->schema->has($inputName)) {
            // Get the element and force set the property
            $element = $this->schema->get($inputName);
            $element['form'][$property] = $value;

            // Push back the modifyed element in the schema
            $this->schema->set($inputName, $element);
        }

        return $this;
    }

    /**
     * Function used to set options of a select element. Shortcut for using
     * `setInputArgument` and `setValue`.
     *
     * @param string               $inputName The select name to add options to
     * @param array<string,string> $data      An array of `value => label` options
     * @param string               $selected  The selected key
     *
     * @return self
     */
    public function setOptions(string $inputName, array $data = [], ?string $selected = null)
    {
        // Set options
        $this->setInputArgument($inputName, 'options', $data);

        // Set the value
        if (!is_null($selected)) {
            $this->setValue($inputName, $selected);
        }

        return $this;
    }

    /**
     * Function to set the form namespace.
     * Use the form namespace to wrap the fields name in a top level array.
     * Useful when using multiple schemas at once or if the names are using dot syntaxt.
     * See : http://stackoverflow.com/a/20365198/445757.
     *
     * @param string $namespace
     *
     * @return self
     */
    public function setFormNamespace(string $namespace)
    {
        $this->formNamespace = $namespace;

        return $this;
    }

    /**
     * Get default input type for element without one.
     *
     * @return string
     */
    public function getDefaultType(): string
    {
        return $this->defaultType;
    }

    /**
     * Set default input type for element without one.
     *
     * @param string $defaultType Default type name (eg. select, text, phone, mytext, etc.)
     *
     * @return self
     */
    public function setDefaultType(string $defaultType)
    {
        $this->defaultType = $defaultType;

        return $this;
    }

    /**
     * Generate an array contining all nececerry value to generate a form
     * with Twig.
     *
     * @return array<string,string> The form fields data
     */
    public function generate(): array
    {
        $form = collect([]);

        // Loop all the the fields in the schema
        foreach ($this->schema->all() as $name => $input) {
            // Skip if it doesn't have a `form` definition
            if (!isset($input['form'])) {
                continue;
            }

            // Alias the schema input
            $element = $input['form'];

            // Get the value from the data
            $value = $this->getValueForName($name);

            // Add the namespace to the name if it's defined
            $name = $this->getNamespacedName($name);

            // If element doesn't have a type, inject default (text)
            if (!isset($element['type'])) {
                $element['type'] = $this->defaultType;
            }

            // Get the class FQN for the given type.
            // Fallback to default if not found
            try {
                $elementClass = $this->getType($element['type']);
            } catch (InputNotFoundException $e) {
                $elementClass = $this->getType($this->getDefaultType());
            }

            // Create the actual class
            $class = new $elementClass($name, $element, $value);

            // Make sure we have an instance of InputInterface.
            // An error will be thrown otherwise
            if (!$class instanceof InputInterface) {
                throw new InvalidClassException('Input class needs to implement InputInterface');
            }

            // Put parsed data from the InputInterface to `$form`
            // Collection put is used here, because of the namespace, which will put it at the right place
            $form->put($name, $class->parse());
        }

        return $form->toArray();
    }

    /**
     * Gets the defined value for the specified field name.
     *
     * @param string $name
     *
     * @return mixed
     */
    protected function getValueForName(string $name): mixed
    {
        return Arr::get($this->data, $name);
    }

    /**
     * Get the name of the field, wrapped by the global form namespace.
     *
     * @param string $name
     *
     * @return string
     */
    protected function getNamespacedName(string $name): string
    {
        if ($this->formNamespace !== '') {
            return $this->formNamespace . '[' . $name . ']';
        }

        return $name;
    }

    /**
     * Register the default built-in types.
     */
    protected function registerDefaultType(): void
    {
        $this->registerType('alert', Alert::class);
        $this->registerType('checkbox', Checkbox::class);
        $this->registerType('hidden', Hidden::class);
        $this->registerType('select', Select::class);
        $this->registerType('text', Text::class);
        $this->registerType('textarea', Textarea::class);
    }
}
